apps/emqx/src/bpapi/emqx_bpapi_static_checks.erl

```diff
@@ -0,0 +1,147 @@
+    Callers = find_remote_calls(Opts),
+    {BPAPICalls, NonBPAPICalls} = lists:partition(fun is_bpapi_call/1, Callers),
+    warn_nonbpapi_rpcs(NonBPAPICalls),
+    CombinedAPI = collect_bpapis(BPAPICalls, PLT),
+    dump_api(CombinedAPI),
+    erase(bpapi_ok).
+
+prepare(#{reldir := RelDir, plt := PLT}) ->
+    ?INFO("Starting xref...", []),
+    xref:start(?XREF),
+    filelib:wildcard(RelDir ++ "/*/ebin/") =:= [] andalso
+        error("No applications found in the release directory. Wrong directory?"),
+    xref:set_default(?XREF, [{warnings, false}]),
+    xref:add_release(?XREF, RelDir),
+    %% Now to the dialyzer stuff:
+    ?INFO("Loading PLT...", []),
+    dialyzer_plt:from_file(PLT).
+
+find_remote_calls(_Opts) ->
+    Query = "XC | (A - [" ?IGNORED_APPS "]:App)
+               || ([" ?RPC_MODULES "] : Mod - " ?IGNORED_RPC_CALLS ")",
+    {ok, Calls} = xref:q(?XREF, Query),
+    ?INFO("Calls to RPC modules ~p", [Calls]),
+    {Callers, _Callees} = lists:unzip(Calls),
+    Callers.
+
+-spec warn_nonbpapi_rpcs([mfa()]) -> ok.
+warn_nonbpapi_rpcs([]) ->
+    ok;
+warn_nonbpapi_rpcs(L) ->
+    setnok(),
+    lists:foreach(fun({M, F, A}) ->
+                          ?ERROR("~p:~p/~p does a remote call outside of a dedicated "
+                                 "backplane API module. "
+                                 "It may break during rolling cluster upgrade",
+                                 [M, F, A])
+                  end,
+                  L).
+
+-spec is_bpapi_call(mfa()) -> boolean().
+is_bpapi_call({Module, _Function, _Arity}) ->
+    case catch Module:bpapi_meta() of
+        #{api := _} -> true;
+        _           -> false
+    end.
+
+-spec dump_api(fulldump()) -> ok.
+dump_api(Term) ->
+    Filename = filename:join(code:priv_dir(emqx_bpapi), emqx_app:get_release() ++ ".bpapi"),
+    file:write_file(Filename, io_lib:format("~0p.", [Term])).
+
+-spec collect_bpapis([mfa()], _PLT) -> fulldump().
+collect_bpapis(L, PLT) ->
+    Modules = lists:usort([M || {M, _F, _A} <- L]),
+    lists:foldl(fun(Mod, Acc) ->
+                        #{ api     := API
+                         , version := Vsn
+                         , calls   := Calls0
+                         , casts   := Casts0
+                         } = Mod:bpapi_meta(),
+                        Calls = enrich(PLT, Calls0),
+                        Casts = enrich(PLT, Casts0),
+                        Acc#{API => #{Vsn => #{ calls => Calls
+                                              , casts => Casts
+                                              }}}
+                end,
+                #{},
+                Modules
+               ).
+
+%% Add information about types from the PLT
+-spec enrich(_PLT, [emqx_bpapi:rpc()]) -> [typed_rpc()].
+enrich(PLT, Calls) ->
+    [case {lookup_type(PLT, From), lookup_type(PLT, To)} of
+         {{value, TFrom}, {value, TTo}} ->
+             {{From, TFrom}, {To, TTo}};
+         {_, none} ->
```

Comment:
Q: could it happen that the typespecs of `From` could be missing?

Reply:
This could happen if PLT is wrong or out-of-date. I guess it should just crash with badmatch here